--- a/config.def.h	2021-06-30 16:32:20.331922538 +0300
+++ b/config.def.h	2021-06-30 19:09:27.342827269 +0300
@@ -5,11 +5,13 @@
  *
  * font: see http://freedesktop.org/software/fontconfig/fontconfig-user.html
  */
-static char *font = "Liberation Mono:pixelsize=12:antialias=true:autohint=true";
+static char *font = "mono:pixelsize=14:antialias=true:autohint=true";
 /* Spare fonts */
 static char *font2[] = {
-/*	"Inconsolata for Powerline:pixelsize=12:antialias=true:autohint=true", */
-/*	"Hack Nerd Font Mono:pixelsize=11:antialias=true:autohint=true", */
+	"mono:pixelsize=14:antialias=true:autohint=true",
+	"monospace:pixelsize=14:antialias=true:autohint=true",
+	"JoyPixels:pixelsize=10:antialias=true:autohint=true",
+	"FontAwesome:pixelsize=10:antialias=true:autohint=true",
 };
 
 static int borderpx = 2;
