diff -up ../st-0.8.4/st.h ./st.h
--- ../st-0.8.4/st.h	2020-06-19 12:29:45.000000000 +0300
+++ ./st.h	2021-06-29 16:02:21.653740131 +0300
@@ -81,6 +81,9 @@ void die(const char *, ...);
 void redraw(void);
 void draw(void);
 
+void externalpipe(const Arg *);
+void kscrolldown(const Arg *);
+void kscrollup(const Arg *);
 void printscreen(const Arg *);
 void printsel(const Arg *);
 void sendbreak(const Arg *);
@@ -123,3 +124,5 @@ extern char *termname;
 extern unsigned int tabspaces;
 extern unsigned int defaultfg;
 extern unsigned int defaultbg;
+extern float alpha, alphaUnfocused;
+extern const int boxdraw, boxdraw_bold, boxdraw_braille;
